server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -486,15 +499,16 @@ private static void parseObjectDynamic(DocumentParserContext context, String cur
                     );
                 }
                 if (dynamicObjectMapper instanceof ObjectMapper) {
-                    throw new DocumentParsingException(
-                        context.parser().getTokenLocation(),
-                        "Tried to add subobject ["
-                            + dynamicObjectMapper.simpleName()
-                            + "] to object ["
-                            + context.parent().name()
-                            + "] which does not support subobjects"
-                    );
+                    // We have an ObjectMapper but subobjects are disallowed
+                    // therefore we create a new DocumentParserContext that
+                    // prepends currentFieldName to any immediate children.
+                    parseObjectOrNested(context.createFlattenContext(currentFieldName));
+                    return;
                 }
+
+            }
+            if (context.dynamic() != ObjectMapper.Dynamic.RUNTIME) {
```

Comment:
nit: for readability, should this have been if dynamic == Dynamic.TRUE ? Otherwise you have to look above to see that we can only have either RUNTIME or TRUE at this point.

Reply:
After https://github.com/elastic/elasticsearch/pull/96235 is merged, this case should also cover `Dynamic.UNTIL_LIMIT`, which it currently does.

Reply:
I'd still prefer to be explicit, rather than assuming that any other value than RUNTIME should get the dynamic behaviour. Still it's not a super huge deal, but I think it'd improve readability

Reply:
> I assume this came from refactoring out context.addDynamicMapper(dynamicObjectMapper)

Yes this was the case, but I totally agree with @javanna on the readability aspect.